tests/ConnectDisconnectTest.test.nut

```diff
@@ -0,0 +1,211 @@
+        assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        resolve();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+      assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(cm.isConnected(), "CM should report state as connected!");
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+  /*
+  * Verifies that more than one disconnect() invocation doesn't result in incorect behavior
+  *
+  */
+  function testDoubleDisconnectAsync() {
+    assertTrue(server.isconnected(), "Failed pre-run saanity check");
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!server.isconnected(), "should NOT be connected!");
+        assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+        cm.connect();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+      assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+
+      cm.onConnect(function() {
+        resolve();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+      assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(cm.isConnected(), "CM should report state as connected!");
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+/*
+* Verifies that more than one connect() invocation doesn't lead to disconnection
+*
+*/
+  function testDoubleConnectAsync() {
+    assertTrue(server.isconnected(), "Failed pre-run saanity check");
+    return Promise(function(resolve, reject) {
+      cm.connect();
+      assertTrue(server.isconnected(), "should be connected!");
+      assertTrue(cm.isConnected(), "CM should report connected state!");
+
+      cm.onConnect(function() {
+        assertTrue(server.isconnected(), "should be connected!");
+        assertTrue(cm.isConnected(), "CM should report connected state!");
+        resolve();
+      }.bindenv(this));
+
+      cm.connect();
+      assertTrue(server.isconnected(), "should be connected!");
+      assertTrue(cm.isConnected(), "CM should report connected state!");
+    }.bindenv(this))
+    .then(function(val) {
```

Comment:
lines 124-133 can be reused across the code. you can fin the same code in most of (if not all) tests